jmclient/jmclient/wallet_rpc.py

```diff
@@ -0,0 +1,659 @@
+        wss_factory is a twisted protocol factory for the
+        websocket connections for clients to subscribe to updates.
+        """
+        # cookie tracks single user's state.
+        self.cookie = None
+        self.port = port
+        self.wss_port = wss_port
+        self.tls = tls
+        pref = "wss" if self.tls else "ws"
+        self.wss_url = pref + "://127.0.0.1:" + str(wss_port)
+        # the collection of services which this
+        # daemon may switch on and off:
+        self.services = {}
+        # master single wallet service which we
+        # allow the client to start/stop.
+        self.services["wallet"] = None
+        self.wallet_name = "None"
+        # label for convenience:
+        self.wallet_service = self.services["wallet"]
+        # Client may start other services, but only
+        # one instance.
+        self.services["snicker"] = None
+        self.services["maker"] = None
+        # our taker object will handle doing sends/taker-cjs:
+        self.taker = None
+        # the factory of type JmwalletdWebsocketServerFactory,
+        # which has notification methods that can be passed
+        # as callbacks for in-wallet events:
+        self.wss_factory = None
+        # ensure shut down does not leave dangling services:
+        atexit.register(self.stopService)
+
+    def startService(self):
+        """ Encapsulates start up actions.
+        Here starting the TLS server.
+        """
+        super().startService()
+        # we do not auto-start any service, including the base
+        # wallet service, since the client must actively request
+        # that with the appropriate credential (password).
+        # initialise the web socket service for subscriptions
+        self.wss_factory = JmwalletdWebSocketServerFactory(self.wss_url)
+        self.wss_factory.protocol = JmwalletdWebSocketServerProtocol
+        if self.tls:
+            cf = get_ssl_context(os.path.join(jm_single().datadir, "ssl"))
+            listener_rpc = reactor.listenSSL(self.port, Site(
+                self.app.resource()), contextFactory=cf)
+            listener_ws = listenWS(self.wss_factory, contextFactory=cf)
+        else:
+            listener_rpc = reactor.listenTCP(self.port, Site(
+                self.app.resource()))
+            listener_ws = listenWS(self.wss_factory, contextFactory=None)
+        return (listener_rpc, listener_ws)
+
+    def stopService(self):
+        """ Encapsulates shut down actions.
+        """
+        # Currently valid authorization tokens must be removed
+        # from the daemon:
+        self.cookie = None
+        # if the wallet-daemon is shut down, all services
+        # it encapsulates must also be shut down.
+        for name, service in self.services.items():
+            if service:
+                service.stopService()
+        super().stopService()
+
+    @app.handle_errors(NotAuthorized)
+    def not_authorized(self, request, failure):
```

Comment:
All these `failure` parameters are not used, do they have a purpose?

Reply:
About this one I'll have to check; it's part of Klein's syntax; I presume this gets called back somewhere.

Reply:
Checked, indeed the Klein decorator `handle_errors` passes the handler a request and a failure object (at least in this default syntax ... see `Klein._app.py` for more). Seems fine for now.